src/ballet/http/fd_http_server.c

```diff
@@ -213,21 +204,34 @@ fd_http_server_listen( fd_http_server_t * http,
   http->socket_fd = sockfd;
   http->pollfds[ http->max_conns+http->max_ws_conns ].fd = http->socket_fd;
 
+  FD_LOG_NOTICE(( "listening to port %hu", port ));
+
   return http;
 }
 
 static void
 close_conn( fd_http_server_t * http,
             ulong              conn_idx,
             int                reason ) {
+  if( http->pollfds[ conn_idx ].fd == -1 ) return;
 #ifdef FD_HTTP_SERVER_DEBUG
   FD_LOG_NOTICE(( "Closing connection %lu (fd=%d) (%d-%s)", conn_idx, http->pollfds[ conn_idx ].fd, reason, fd_http_server_connection_close_reason_str( reason ) ));
 #endif
   if( FD_UNLIKELY( -1==close( http->pollfds[ conn_idx ].fd ) ) ) FD_LOG_ERR(( "close failed (%i-%s)", errno, strerror( errno ) ));
   http->pollfds[ conn_idx ].fd = -1;
+  http->pollfds[ conn_idx ].events &= ~POLLOUT;
   if( FD_LIKELY( conn_idx<http->max_conns ) ) {
-    if( FD_LIKELY( http->callbacks.close    ) ) http->callbacks.close( conn_idx, reason, http->callback_ctx );
+    struct fd_http_server_connection * conn = &http->conns[ conn_idx ];
+    if( FD_LIKELY( http->callbacks.close    ) ) http->callbacks.close( conn_idx, reason, &conn->response, http->callback_ctx );
+
   } else {
+    struct fd_http_server_ws_connection * conn = &http->ws_conns[ conn_idx-http->max_conns ];
+    while( conn->send_frame_cnt ) {
+      fd_http_server_ws_frame_t * frame = &conn->send_frames[ conn->send_frame_idx ];
+      if( FD_LIKELY( http->callbacks.ws_sent ) ) http->callbacks.ws_sent( conn_idx-http->max_conns, frame, http->callback_ctx );
```

Comment:
This isn't `ws_sent`? The frame was dropped due to connection close

Reply:
Changed to ws_done